python/plugins/processing/algs/qgis/KNearestConcaveHull.py

```diff
@@ -0,0 +1,526 @@
+        self.addParameter(QgsProcessingParameterNumber(self.KNEIGHBORS,
+                                                       self.tr('Number of neighboring points to consider (a lower number is more concave, a higher number is smoother)'),
+                                                       QgsProcessingParameterNumber.Integer,
+                                                       defaultValue=3, minValue=3))
+        self.addParameter(QgsProcessingParameterField(self.FIELD,
+                                                      self.tr('Field (set if creating concave hulls by class)'),
+                                                      parentLayerParameterName=self.INPUT, optional=True))
+        self.addParameter(QgsProcessingParameterFeatureSink(self.OUTPUT, self.tr('Concave hull'),
+                                                            QgsProcessing.TypeVectorPolygon))
+
+    def processAlgorithm(self, parameters, context, feedback):
+        # get variables from dialog
+        source = self.parameterAsSource(parameters, self.INPUT, context)
+        if source is None:
+            raise QgsProcessingException(self.invalidSourceError(parameters, self.INPUT))
+
+        field_name = self.parameterAsString(parameters, self.FIELD, context)
+        kneighbors = self.parameterAsInt(parameters, self.KNEIGHBORS, context)
+        use_field = bool(field_name)
+
+        field_index = -1
+
+        fields = QgsFields()
+        fields.append(QgsField('id', QVariant.Int, '', 20))
+
+        # get properties of the field the grouping is based on
+        if use_field:
+            field_index = source.fields().lookupField(field_name)
+            if field_index >= 0:
+                fields.append(source.fields()[field_index])
+
+        # initialize writer
+        (sink, dest_id) = self.parameterAsSink(parameters, self.OUTPUT, context,
+                                               fields, QgsWkbTypes.Polygon, source.sourceCrs())
+
+        current = 0
+        fid = 0
+
+        if field_index >= 0:
+            # get unique values of field denoted by index as filter conditions
+            unique_values = source.uniqueValues(field_index)
+            total = 100.0 / float(source.featureCount() * len(unique_values))
+
+            for unique in unique_values:
+                points = []
+                filter = QgsExpression.createFieldEqualityExpression(field_name, unique)
+                request = QgsFeatureRequest().setFilterExpression(filter)
+                request.setSubsetOfAttributes([])
+                features = source.getFeatures(request)
+                for in_feature in features:
+                    points.extend(extract_points(in_feature.geometry()))
+                    current += 1
+                    feedback.setProgress(int(current * total))
+
+                # A minimum of 3 points is necessary to proceed
+                if len(points) >= 3:
```

Comment:
Hi @rudivs is not clear to me why this check "if len(points) >= 3" but not also at line 188.
Another point is that id len(points) < 3 the algorithm return an empty sink... is this the intended behaviour?

Reply:
Both good points. It's not strictly necessary because the concave hull algorithm does its own checks, but this makes it easier to catch a clear error scenario early. I've added the check to the case when there is no grouping field. Will push through some commits shortly.